### PR TITLE
[#593] Add smartparens-clojure.el

### PR DESCRIPTION
**Pull request by vspinu**
Make 'sp-sexp-prefix' aware of Clojure constructs.

### CODE REVIEW

**Comment by vspinu** [comment]
To be clear about the issues. The `(sp-wrap-with-pair "(")` works as expected on the following:

```clojure
#{...}
^{:a 1}
#(xyzzy 1 2 %)
#_(comment reader macro)
(#?(:clj     Double/NaN
    :cljs    js/NaN
    :default nil))
~unquote-me
`~'unquote-me
~@(splice-me)
`(quote-me)
```quote-me
@(deref-me)
@deref-me
#?@(:clj [3 4] :cljs [5 6])
```

but fails on 

```clojure
#"..."
#'xyzzy
```

`sp-forward-barf-sexp` has issues with all prefixes that include `#`. For example `(|#{...})` and `sp-forward-barf-sexp` creates `(#){...}` instead of `()#{...}`. Still better than nothing.

File: smartparens-clojure.el [hunk 1]
```diff
@@ -0,0 +1,45 @@
+;;; smartparens-clojure.el --- Additional configuration for Clojure mode.
+;;
+;; Author: Matus Goljer <matus.goljer@gmail.com>
```
Comment (Fuco1) [inline smartparens-clojure.el]: You are the author, not me.

File: smartparens-clojure.el [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+;; Maintainer: Matus Goljer <matus.goljer@gmail.com>
+;; Created: 8 February 2015
```
Comment (Fuco1) [inline smartparens-clojure.el]: Date

**Comment by Fuco1** [comment]
It would be most awesomest if you could add an ert test for the examples you gave above.

Here's a template you can use

```elisp
(ert-deftest sp-test-somethingsomething-clojure ()
  (sp-test-with-temp-buffer "initial buffer content, | is the initial point"
      (clojure-mode)
    (your-code-in-a-new-temp-clojure-buffer)
    (optionally (insert "|") so you can test the resulting point)
    (should (equal (buffer-string) "output | buffer"))))
```

See the python tests for more inspiration.

**Comment by Fuco1** [comment]
@vspinu You should probably open a separate issue about the barf behaviour, I smell a general problem there.  Thanks!

File: smartparens-clojure.el (hunk 1 above)
Comment (vspinu) [inline smartparens-clojure.el]: > You are the author, not me.

This is a trivial copy paste change of your code and suggestions. I am a bit wary of 'Maintainer' field as I am a somewhat intermitent clojure user and might not hold to the promise. 

In any case I can un-modestly put my name there if you prefer so.

Comment (Fuco1) [inline smartparens-clojure.el]: You can leave me as the maintainer :P

Comment (Fuco1) [inline smartparens-clojure.el]: It's just for the legal nonsense in the case we eventually get into emacs proper... other authors will be added as they contribute to the file.

**Comment by vspinu** [comment]
>  could add an ert test 

Sure. I will add ert tests. 

I see some corner cases with the current regexp. I would like to dig a bit more into smartparens inner workings first.

**Comment by Fuco1** [comment]
Hey, just checking in, any progress? :)

**Comment by vspinu** [comment]
Unfortunately not.  Sorry about this. I will try to get this rolling by the end of the week.

**Comment by Fuco1** [comment]
No problem :)

**Comment by vspinu** [comment]
I think I am done  for now. A bunch of tests are commented out as they break because of #621. 

Unfortunately by simply setting `#` to prefix syntax doesn't solve all issues. If the prefix ends with a  symbol characters  (like in `#?(....)`)  the operations will still fail. 

 I don't know what would be the proper fix for this. An alternative to `sp-sexp-regexp` might be to introduce a special syntax table that would be active during sp commands.

**Comment by Fuco1** [comment]
Hm, that is odd, I think the regexp shouldn't care about syntax properties.

Anyhow, let's merge this and I'll add you as a contributor so you can take care of the clojure stuff if you find some more issues.  You can also try to work on 621 but that seems more of an internal issue, so don't feel pressured :)

Thanks for the contribution.
